[AArch64][PAC] Lower ptrauth constants in code for MachO.

ahmedbougacha opened the pull request:
This also adds support for auth stubs on MachO using __DATA,__auth_ptr.

Some of the machinery for auth stubs is already implemented;  this
generalizes that a bit to support MachO, and moves some of the shared logic into MMIImpls.
In particular, this originally had an AuthStubInfo struct, but we no longer need it
beyond a single MCExpr.  So this provides variants of the symbol stub
helper type declarations and functions for "expr stubs", where a
stub points at an arbitrary MCExpr, rather than a simple MCSymbol (and a bit).

On MachO, the auth stubs are emitted in __DATA,__auth_ptr.

Beyond that, this is mostly straightforward.

ahmedbougacha committed:
[AArch64][PAC] Lower ptrauth constants in code for MachO.

ahmedbougacha committed:
[AArch64][PAC] Support auth stubs on MachO using __DATA,__auth_ptr.

Some of the machinery for auth stubs is already implemented;  this
generalizes that a bit to support MachO.

This also moves some of the shared logic into MMIImpls.  In particular,
this originally had an AuthStubInfo struct, but we no longer need it
beyond a single MCExpr.  So this provides variants of the symbol stub
helper type declarations and functions for "expr stubs", where a
stub points at an arbitrary MCExpr, rather than a simple MCSymbol.

On MachO, the auth stubs are emitted in __DATA,__auth_ptr.

kovdan01 commented on llvm/lib/Target/AArch64/AArch64AsmPrinter.cpp:
```diff
@@ -848,20 +848,39 @@ void AArch64AsmPrinter::emitHwasanMemaccessSymbols(Module &M) {
 
   const Triple &TT = TM.getTargetTriple();
   if (TT.isOSBinFormatMachO()) {
+
```
This empty new line was probably inserted unintentionally.

kovdan01 commented on llvm/lib/Target/AArch64/AArch64AsmPrinter.cpp:
```diff
@@ -848,20 +848,39 @@ void AArch64AsmPrinter::emitHwasanMemaccessSymbols(Module &M) {
+                                     SectionKind::getMetadata()));
+      emitAlignment(Align(8));
+
+      for (auto &Stub : Stubs)
```
```suggestion
      for (const auto &Stub : Stubs)
```

kovdan01 commented on llvm/lib/CodeGen/MachineModuleInfoImpls.cpp:
```diff
@@ -43,24 +43,19 @@ MachineModuleInfoImpl::SymbolListTy MachineModuleInfoImpl::getSortedStubs(
-  typename MachineModuleInfoTarget::AuthStubListTy List(AuthPtrStubs.begin(),
-                                                        AuthPtrStubs.end());
+using ExprStubPairTy = std::pair<MCSymbol *, const MCExpr *>;
+static int SortAuthStubPair(const ExprStubPairTy *LHS,
```
It's probably worth adding a test which checks that stubs are sorted. I'm OK with merging this "as is" and implementing new tests as a separate patch later.

ahmedbougacha replied:
Yeah, the ordering doesn't matter as long as it's stable;  I added another extern_weak as the barest of sanity checks.

kovdan01 reviewed (approved):
LGTM with a couple of minor nits. Thanks for getting rid of useless `struct AuthStubInfo` in favor of just `const MCExpr *`.

kovdan01 commented on llvm/include/llvm/CodeGen/MachineModuleInfoImpls.h:
```diff
@@ -36,6 +36,11 @@ class MachineModuleInfoMachO : public MachineModuleInfoImpl {
   DenseMap<MCSymbol *, StubValueTy> ThreadLocalGVStubs;
 
+  /// Darwin '$auth_ptr' stubs.  The key is the stub symbol, like
+  /// "Lfoo$addend$auth_ptr$ib$12".  The value is the MCExpr representing that
```
Just in case you've missed that - currently, there is no addend support for `$auth_ptr$` symbols - see `getAuthPtrSlotSymbolHelper`. It's probably worth mentioning that addend is optional in this comment. Feel free to ignore though.

ahmedbougacha replied:
Right, I don't think we'll ever need the addend support here, I'm keeping it downstream for some other use-cases.

I should mention we're also missing the address-diversity support here, and that one we'll likely want to add here in the future, at least for MachO;  we'll see then whether it's needed for ELF, given the broader support for auth GOT relocations that MachO doesn't expose.

ahmedbougacha committed:
Address review comments.

Move the extern_weak testcase to be adjacent to the emitted stubs,
and add another testcase to check emission stability.
